### PR TITLE
App config constance

### PR DESCRIPTION
**Pull request by smk4664**
# Closes: #3403
# What's Changed

Provide support for configuring Apps through the Nautobot Settings Config. Updates the settings_or_config templatetag helper and jinja2 filter to add App support.

Screenshot 2023-03-15 at 11 57 16 PM
Screenshot 2023-03-06 at 9 52 16 PM

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Example Plugin Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

**Review by glennmatthews — COMMENTED** [review]
Cool!

File: nautobot/docs/plugins/development.md [hunk 1]
```diff
@@ -162,15 +162,16 @@ Nautobot looks for the `config` variable within an app's `__init__.py` to load i
 | `min_version` | `None` | Minimum version of Nautobot with which the app is compatible |
 | `required_settings` | `[]` | A list of any configuration parameters that **must** be defined by the user |
 | `searchable_models` | `[]` | A list of model names to include in the global Nautobot search |
+| `constance_config` | `{}` | A [Django Constance](#adding-database-backed-config) configuration parameter for settings.
```
Comment (glennmatthews) [inline nautobot/docs/plugins/development.md:165]:
```suggestion
| `constance_config` | `{}` | [Django Constance](#adding-database-backed-config) configuration parameters for settings. |
```

**Review by glennmatthews — COMMENTED** [review]
Looks really solid! The only thing I can think might be missing is whether the constance_configs should be added to `self.features` and shown on the "installed plugins" detail view? Doesn't look like we're currently doing that for the `settings` configs, but it might be useful to add? If so, I'm thinking it probably should only just list the variable names/help texts/field types, **not** the currently configured values?

**Comment by smk4664** [comment]
> Looks really solid! The only thing I can think might be missing is whether the constance_configs should be added to `self.features` and shown on the "installed plugins" detail view? Doesn't look like we're currently doing that for the `settings` configs, but it might be useful to add? If so, I'm thinking it probably should only just list the variable names/help texts/field types, **not** the currently configured values?

That is an interesting idea. Let me look into that.

File: nautobot/extras/plugins/utils.py [hunk 2]
```diff
@@ -107,7 +108,16 @@ def load_plugin(plugin_name, settings):
...
     # Update Constance Config and Constance Fieldset
     if plugin_config.constance_config:
-        app_config = {f"{plugin_name}__{key}": value for key, value in plugin_config.constance_config.items()}
+        app_config = {}
+        for key, value in plugin_config.constance_config.items():
+            config_item = value
+            # Enforce ConstanceConfigItem namedtuple
+            if not isinstance(value, ConstanceConfigItem):
+                config_item = ConstanceConfigItem(*value)
+                plugin_config.constance_config[key] = config_item
```
Comment (glennmatthews) [inline nautobot/extras/plugins/utils.py:117]: Do we actually read from `plugin_config.constance_config` ever again after this point? In other words, is this actually necessary/useful?

File: nautobot/apps/config.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+"""Helper code for loading values that may be defined in settings.py/nautobot_config.py *or* in django-constance."""
+
+from constance import config
+from django.conf import settings
+
```
Comment (glennmatthews) [inline nautobot/apps/config.py:5]: Probably should expose `ConstanceConfigItem` class in the `nautobot.apps` namespace, either here in `config.py` or else in `__init__.py` alongside `NautobotAppConfig`.

File: examples/example_plugin/example_plugin/__init__.py [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 
 from nautobot.core.signals import nautobot_database_ready
+from nautobot.core.settings_funcs import ConstanceConfigItem
 from nautobot.apps import NautobotAppConfig
```
Comment (glennmatthews) [inline examples/example_plugin/example_plugin/__init__.py:12]:
```suggestion
from nautobot.apps import ConstanceConfigItem, NautobotAppConfig, nautobot_database_ready
```

File: nautobot/core/templates/admin/config/config.html [hunk 5]
```diff
@@ -42,32 +42,65 @@
...
-                                        <div class="alert alert-warning" role="alert">
-                                            <code>{{ item.name }} = {{ settings_value | quote_string }}</code>
-                                            is defined in <code>{{ settings.SETTINGS_PATH }}</code>,
-                                            therefore it is not currently editable here.
-                                            <br>Remove that declaration from <code>{{ settings.SETTINGS_PATH }}</code>
-                                            if you want to manage it via this interface instead.
+                                {# Handle App Configuration #}
+                                {% if "__" in item.name %}
+                                    {% with app_name=item.name|split:"__"|first name=item.name|split:"__"|last %}
+                                        {% get_attr settings "PLUGINS_CONFIG" as app_config %}
```
Comment (glennmatthews) [inline nautobot/core/templates/admin/config/config.html:48]: Perhaps change `app_config` to `app_settings` throughout this file for clarity?

File: nautobot/core/templates/admin/config/config.html [hunk 6]
```diff
@@ -42,32 +42,65 @@
...
+                                                    is defined in <code>{{ settings.SETTINGS_PATH }}</code>
+                                                    under PLUGINS_CONFIG[{{app_name}}],
+                                                    therefore it is not currently editable here.
+                                                    <br>Remove that declaration from <code>{{ settings.SETTINGS_PATH }}</code>
+                                                    if you want to manage it via this interface instead.
+                                                </div>
+                                                {{ item.form_field.as_hidden }}
+                                            {% else %}
+                                                <label class="col-md-3 control-label" for="{{ item.form_field.id_for_label }}">
+                                                    {{ name | split:"_" | join:" " }}
```
[the remark this answers is not in the document]
Comment (smk4664) [reply nautobot/core/templates/admin/config/config.html:62]: Each App will have their own Fieldset. I don't believe this is necessary.

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/admin/config/config.html:62]: Is there any labeling of each Fieldset to visually indicate which app it belongs to? I'm not seeing that in the template.

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/templates/admin/config/config.html:62]: Or is that automagically coming in as `fieldset.title`?

[the remark this answers is not in the document]
Comment (smk4664) [reply nautobot/core/templates/admin/config/config.html:62]: Yep, coming in from fieldset.title. When the plugin is registered in `nautobot.extras.plugins.utils.load_plugin` I am using the verbose_name as the title and adding the keys from the `plugin_config.constance_config` into the fieldset.

File: nautobot/apps/config.py (hunk 3 above)
Comment (smk4664) [reply nautobot/apps/config.py:5]: Yep, added.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**